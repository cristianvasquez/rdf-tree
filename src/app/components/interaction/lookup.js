import rdf from 'rdf-ext'

function getRelatedTerms (dataset, term) {

  // Track all graphs encountered
  const allGraphs = rdf.termSet()

  // Track which graphs each ID appears in
  const termToGraphs = rdf.termMap()

  // Helper to record ID-graph relationships
  const recordGraphs = (term, { graph }) => {
    // if (graph?.value) {
      allGraphs.add(graph)
      if (!termToGraphs.has(term)) {
        termToGraphs.set(term, rdf.termSet())
      }
      termToGraphs.get(term).add(graph)
    // }
  }

  const inTerms = rdf.termSet()
  for (const quad of dataset.match(null, null, term)) {
    inTerms.add(quad.subject)
    recordGraphs(quad.subject, quad)
  }
  const outTerms = rdf.termSet()
  for (const quad of dataset.match(term, null, null)) {
    outTerms.add(quad.object)
    recordGraphs(quad.object, quad)
  }

  termToGraphs.set(term, allGraphs)

  return {
    incomingTerms: [...inTerms],
    outgoingTerms: [...outTerms],
    graphs: Array.from(allGraphs), // All found graphs
    termToGraphs, // Map of term -> Set of graphs Set
  }
}

export { getRelatedTerms }
