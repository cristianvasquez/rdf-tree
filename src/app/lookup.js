import rdf from 'rdf-ext'

function getRelatedIds (dataset, term, uriToIds) {
  const toIds = (term) => uriToIds.has(term) ? [...uriToIds.get(term)] : []

  // Track all graphs encountered
  const allGraphs = new Set()

  // Track which graphs each ID appears in
  const idToGraphs = new Map()

  // Helper to record ID-graph relationships
  const recordGraphs = (id, quad) => {
    const graph = quad.graph.value || 'default'
    allGraphs.add(graph)

    if (!idToGraphs.has(id)) {
      idToGraphs.set(id, new Set())
    }
    idToGraphs.get(id).add(graph)
  }

  // Process incoming relations
  const inTerms = rdf.termSet()
  const incoming = Array.from(new Set(
    [...dataset.match(null, null, term)].flatMap(quad => {
      const ids = toIds(quad.subject)
      inTerms.add(quad.subject)
      ids.forEach(id => recordGraphs(id, quad))
      return ids
    }),
  ))

  // Process same term
  const same = toIds(term)
  same.forEach(id => {
    // For "same" IDs, we need to look up what graphs they appear in
    const sameQuads = [
      ...dataset.match(term, null, null),
      ...dataset.match(null, null, term)]
    sameQuads.forEach(quad => {
      recordGraphs(id, quad)
    })
  })

  // Process outgoing relations
  const outTerms = rdf.termSet()
  const outgoing = Array.from(new Set(
    [...dataset.match(term, null, null)].flatMap(quad => {
      const ids = toIds(quad.object)
      outTerms.add(quad.object)
      ids.forEach(id => recordGraphs(id, quad))
      return ids
    }),
  ))

  // Convert Sets to Arrays in idToGraphs for easier consumption
  const idToGraphsArray = new Map()
  idToGraphs.forEach((graphs, id) => {
    idToGraphsArray.set(id, Array.from(graphs))
  })

  return {
    incoming,
    same,
    outgoing,
    incomingTerms: [...inTerms],
    outgoingTerms: [...outTerms],
    graphs: Array.from(allGraphs), // All encountered graphs
    idToGraphs: idToGraphsArray,    // Map of id -> graphs
  }
}

export { getRelatedIds }
