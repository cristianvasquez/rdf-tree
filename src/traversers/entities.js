import grapoi from 'grapoi'
import rdf from 'rdf-ext'
import { bfsEntity } from './bfsEntity.js'

const all = { subject: undefined, predicate: undefined, object: undefined }

// Gets the roots of all entities.
//
// matchers: array of matchers of the type {s,p,o}, where {} is all triples
// maxDepth: max depth of the traversal
function getEntities (dataset, options) {

  const { matchers, asTriples, maxDepth } = {
    matchers: [all],
    maxDepth: Infinity,
    ...options,
  }

  const result = []

  // Changed from termSet to termMap for the URI -> ID mapping
  let visited = rdf.termMap()

  for (const { subject, predicate, object, graph } of matchers) {
    const batch = [...dataset.match(subject, predicate, object, graph)].map(
      x => x.subject).filter(subject => !visited.has(subject))

    const batchResult = batch.reduce((acc, term) => {

      if (acc.visited.has(term)) {
        // If it was visited, ignore it from the batch
        return { entities: acc.entities, visited: acc.visited }
      }

      const pointer = grapoi({ dataset, term, factory: rdf })

      const { entity, visited } = bfsEntity(pointer, {
        maxDepth,
        visited: acc.visited,
      })
      return {
        entities: acc.entities.concat(entity),
        visited,
      }
    }, { entities: [], visited })

    result.push(...batchResult.entities)
    visited = batchResult.visited
  }

  // Return both the entities and the visited map for URI -> ID mapping
  return { entities: result, uriToIds: visited }
}

export { getEntities }
