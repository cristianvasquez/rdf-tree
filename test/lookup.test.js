import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { describe, it } from 'mocha'
import rdf from 'rdf-ext'
import { getRelatedIds } from '../src/app/lookup.js'
import { getEntities } from '../src/traversers/entities.js'
import { getRabbitDataset } from './support/dataset.js'

expect.extend({ toMatchSnapshot })

describe('lookup', () => {

  it(`maps URIs to their corresponding IDs`, function () {
    const dataset = getRabbitDataset()
    const { uriToIds } = getEntities(dataset)
    expect(uriToIds).toMatchSnapshot(this)
  })

  it(`gets Related ids for an entity`, function () {
    const dataset = getRabbitDataset()
    const { uriToIds } = getEntities(dataset)
    const term = rdf.namedNode('http://example.org/note/Alice.md')
    const result = getRelatedIds(dataset, term, uriToIds)
    expect(result).toMatchSnapshot(this)
  })
})
