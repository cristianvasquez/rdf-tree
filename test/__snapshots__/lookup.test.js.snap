// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lookup gets Related ids for an entity 1`] = `
Object {
  "graphs": Array [
    "http://example.org/note/WhiteRabbit.md",
    "http://example.org/note/Alice.md",
  ],
  "idToGraphs": Map {
    "132" => Array [
      "http://example.org/note/WhiteRabbit.md",
    ],
    "141" => Array [
      "http://example.org/note/Alice.md",
      "http://example.org/note/WhiteRabbit.md",
    ],
    "129" => Array [
      "http://example.org/note/Alice.md",
    ],
    "149" => Array [
      "http://example.org/note/Alice.md",
    ],
    "150" => Array [
      "http://example.org/note/Alice.md",
    ],
    "154" => Array [
      "http://example.org/note/Alice.md",
    ],
    "155" => Array [
      "http://example.org/note/Alice.md",
    ],
    "151" => Array [
      "http://example.org/note/Alice.md",
    ],
    "140" => Array [
      "http://example.org/note/Alice.md",
    ],
    "152" => Array [
      "http://example.org/note/Alice.md",
    ],
    "164" => Array [
      "http://example.org/note/Alice.md",
    ],
  },
  "incoming": Array [
    "132",
  ],
  "outgoing": Array [
    "129",
    "149",
    "150",
    "154",
    "155",
    "151",
    "140",
    "152",
    "164",
  ],
  "same": Array [
    "141",
  ],
}
`;

exports[`lookup maps URIs to their corresponding IDs 1`] = `
TermMap {
  "index": Map {
    "<http://example.org/note/WhiteRabbit.md>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/note/WhiteRabbit.md",
      },
      "value": Array [
        "91",
      ],
    },
    "<http://pkm-united.org/Note>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://pkm-united.org/Note",
      },
      "value": Array [
        "92",
        "112",
      ],
    },
    "\\"WhiteRabbit\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "WhiteRabbit",
      },
      "value": Array [
        "93",
      ],
    },
    "\\"Conejo blanco\\"@es" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
        },
        "language": "es",
        "termType": "Literal",
        "value": "Conejo blanco",
      },
      "value": Array [
        "94",
      ],
    },
    "_:n3-6" => Object {
      "term": Object {
        "termType": "BlankNode",
        "value": "n3-6",
      },
      "value": Array [
        "95",
      ],
    },
    "<http://pkm-united.org/Block>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://pkm-united.org/Block",
      },
      "value": Array [
        "96",
        "105",
        "116",
      ],
    },
    "<http://example.org/Rabbit>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Rabbit",
      },
      "value": Array [
        "97",
      ],
    },
    "\\"White rabbit\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "White rabbit",
      },
      "value": Array [
        "98",
        "99",
      ],
    },
    "<https://miro.medium.com/max/720/1*HZazTjGg9EBSOoz34IN-tA.jpeg>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "https://miro.medium.com/max/720/1*HZazTjGg9EBSOoz34IN-tA.jpeg",
      },
      "value": Array [
        "100",
      ],
    },
    "_:b3_b4" => Object {
      "term": Object {
        "termType": "BlankNode",
        "value": "b3_b4",
      },
      "value": Array [
        "101",
        "102",
      ],
    },
    "<http://example.org/note/Charlie.md>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/note/Charlie.md",
      },
      "value": Array [
        "103",
        "115",
        "127",
      ],
    },
    "<http://example.org/note/Alice.md>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/note/Alice.md",
      },
      "value": Array [
        "104",
      ],
    },
    "<http://schema.org/Place>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://schema.org/Place",
      },
      "value": Array [
        "106",
      ],
    },
    "\\"Wozenderlands\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "Wozenderlands",
      },
      "value": Array [
        "107",
        "108",
      ],
    },
    "\\"4879\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "4879",
      },
      "value": Array [
        "109",
      ],
    },
    "\\"5 Wonderland Street\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "5 Wonderland Street",
      },
      "value": Array [
        "110",
      ],
    },
    "\\"Charlie\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "Charlie",
      },
      "value": Array [
        "111",
      ],
    },
    "\\"Alice\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "Alice",
      },
      "value": Array [
        "113",
        "117",
        "118",
      ],
    },
    "_:n3-7" => Object {
      "term": Object {
        "termType": "BlankNode",
        "value": "n3-7",
      },
      "value": Array [
        "114",
      ],
    },
    "<https://miro.medium.com/max/1100/1*xupcHn3b0jEFPkjvuH5Pbw.jpeg>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "https://miro.medium.com/max/1100/1*xupcHn3b0jEFPkjvuH5Pbw.jpeg",
      },
      "value": Array [
        "119",
      ],
    },
    "<shared>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "shared",
      },
      "value": Array [
        "120",
      ],
    },
    "<oh-my>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "oh-my",
      },
      "value": Array [
        "121",
      ],
    },
    "<http://example.org/property/loves%20to%20drink%20tea%20with>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/property/loves%20to%20drink%20tea%20with",
      },
      "value": Array [
        "122",
      ],
    },
    "\\"loves to drink tea with\\"" => Object {
      "term": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "loves to drink tea with",
      },
      "value": Array [
        "123",
      ],
    },
    "<http://example.org/s>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/s",
      },
      "value": Array [
        "124",
      ],
    },
    "<http://example.org/Something>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Something",
      },
      "value": Array [
        "125",
      ],
    },
    "<http://example.org/o>" => Object {
      "term": Object {
        "termType": "NamedNode",
        "value": "http://example.org/o",
      },
      "value": Array [
        "126",
      ],
    },
  },
}
`;
